src/vscode-bicep/src/commands/pasteAsBicep.ts

```diff
@@ -0,0 +1,279 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import {
+  env,
+  TextDocumentChangeEvent,
+  TextDocumentChangeReason,
+  Uri,
+  window,
+  workspace,
+  WorkspaceEdit,
+  Range,
+  MessageItem,
+  ConfigurationTarget,
+} from "vscode";
+import { Command } from "./types";
+import { LanguageClient } from "vscode-languageclient/node";
+import { findOrCreateActiveBicepFile } from "./findOrCreateActiveBicepFile";
+import {
+  callWithTelemetryAndErrorHandling,
+  IActionContext,
+} from "@microsoft/vscode-azext-utils";
+import { Disposable } from "../utils";
+import { areEqualIgnoringWhitespace } from "../utils/areEqualIgnoringWhitespace";
+import { getTextAfterFormattingChanges } from "../utils/getTextAfterFormattingChanges";
+import { bicepConfigurationKeys, bicepLanguageId } from "../language/constants";
+import {
+  BicepDecompileForPasteCommandParams,
+  BicepDecompileForPasteCommandResult,
+} from "../language";
+import { OutputChannelManager } from "../utils/OutputChannelManager";
+import { getBicepConfiguration } from "../language/getBicepConfiguration";
+import { SuppressedWarningsManager } from "./SuppressedWarningsManager";
+
+export class PasteAsBicepCommand implements Command {
+  public static readonly id = "bicep.pasteAsBicep";
+  public readonly id = PasteAsBicepCommand.id;
+  public disclaimerShownThisSession = false;
+
+  public constructor(
+    private readonly client: LanguageClient,
+    private readonly outputChannelManager: OutputChannelManager,
+    private readonly suppressedWarningsManager: SuppressedWarningsManager
+  ) {
+    // Nothing to do
+  }
+
+  public async execute(
+    context: IActionContext,
+    documentUri?: Uri
+  ): Promise<void> {
+    documentUri = await findOrCreateActiveBicepFile(
```

Comment:
Will this command be executed outside of the context of an active Bicep file?

I've noticed a huge delay (sometimes 30 seconds plus) when using some of our other functionalities, which I **think** is due to the `workspace.findFiles` API (used in `findOrCreateActiveBicepFile`) being very slow. Any perceptible delay pasting text could provide a frustrating experience.

Reply:
Yes, that would be findFiles.  I would be interested in knowing the scenarios you've seen this (I've never liked our spec'ed behavior here).

However, it should only call findFiles if the command is called through the command palette (probably rare) or via a shortcut key (which I've removed based on previous discussion).

I fought for using the current editor with commands like this if current editor was a bicep file (instead of showing a dropdown) but was overridden by Uche who asked this question at a community meeting (but I wasn't there so don't know if it was properly asked).  Feel free to open an issue to reconsider that decision.

As of right now, though, it should only happen if you choose Paste As Bicep from the command palette (and never via CTRL+V's automatic paste functionality).